src/subtype.c

```diff
@@ -0,0 +1,569 @@
+        if (x == y && !(e->envidx < e->envsz))
+            return 1;
+        return subtype_unionall(x, (jl_unionall_t*)y, e, 1);
+    }
+    if (jl_is_unionall(x))
+        return subtype_unionall(y, (jl_unionall_t*)x, e, 0);
+    if (jl_is_datatype(x) && jl_is_datatype(y)) {
+        if (x == y) return 1;
+        if (y == (jl_value_t*)jl_any_type) return 1;
+        jl_datatype_t *xd = (jl_datatype_t*)x, *yd = (jl_datatype_t*)y;
+        if (jl_is_type_type(xd) && !jl_is_typevar(jl_tparam0(xd)) && jl_typeof(jl_tparam0(xd)) == yd)
+            // TODO this is not strictly correct, but we don't yet have any other way for
+            // e.g. the argument `Int` to match a `::DataType` slot. Most correct would be:
+            // Int isa DataType, Int isa Type{Int}, Type{Int} more specific than DataType,
+            // !(Type{Int} <: DataType), !isleaftype(Type{Int}), because non-DataTypes can
+            // be type-equal to `Int`.
+            return 1;
+        while (xd != jl_any_type && xd->name != yd->name)
+            xd = xd->super;
+        if (xd == (jl_value_t*)jl_any_type) return 0;
+        if (jl_is_tuple_type(xd))
+            return subtype_tuple(xd, yd, e);
+        if (jl_is_vararg_type(xd)) {
+            // Vararg: covariant in first parameter, invariant in second
+            jl_value_t *xp1=jl_tparam0(xd), *xp2=jl_tparam1(xd), *yp1=jl_tparam0(yd), *yp2=jl_tparam1(yd);
+            // in Vararg{T1} <: Vararg{T2}, need to check subtype twice to
+            // simulate the possibility of multiple arguments, which is needed
+            // to handle the concreteness constraint correctly.
+            if (!subtype(xp1, yp1, e)) return 0;
+            if (!subtype(xp1, yp1, e)) return 0;
+            // Vararg{T,N} <: Vararg{T2,N2}; equate N and N2
+            e->invdepth++;
+            int ans = subtype(xp2, yp2, e) && subtype(yp2, xp2, e);
+            e->invdepth--;
+            return ans;
+        }
+        size_t i, np = jl_nparams(xd);
+        int ans = 1;
+        e->invdepth++;
+        for (i=0; i < np; i++) {
+            jl_value_t *xi = jl_tparam(xd, i), *yi = jl_tparam(yd, i);
+            if (!(xi == yi || (subtype(xi, yi, e) && subtype(yi, xi, e)))) {
+                ans = 0; break;
+            }
+        }
+        e->invdepth--;
+        return ans;
+    }
+    if (jl_is_type(y))
+        return x == jl_bottom_type;
+    return x == y || jl_egal(x, y);
+}
+
+static int exists_subtype(jl_value_t *x, jl_value_t *y, jl_stenv_t *e, int8_t anyunions)
+{
+    int exists;
+    for (exists=0; exists <= anyunions; exists++) {
+        if (e->Runions.stacksize > 0)
+            statestack_set(&e->Runions, e->Runions.stacksize-1, exists);
+        e->Lunions.depth = e->Runions.depth = 0;
+        e->Lunions.more = e->Runions.more = 0;
+        int found = subtype(x, y, e);
+        if (e->Lunions.more) {
+            // return up to forall_exists_subtype. the recursion must have this shape:
```

Comment:
does the comment mean that this condition is triggered only under the described circumstances?